### ISSUE TITLE
Columns EndUpdate and duplicate position

### ISSUE THREAD

**Issue by Dzyszla** [JAM-Software/Virtual-TreeView#1228]
If I have 3 columns, last invisible, but set for two latest same position during columns are under updating, freezing on `EndUpdate`:

```delphi
	with VirtualStringTree1.Header.Columns do
	begin
		BeginUpdate;
		Items[0].Position := 0;
		Items[1].Position := 1;
		Items[2].Options := Items[2].Options - [coVisible];
		Items[2].Position := 1;
		EndUpdate;
	end;
```

bug is in the `VirtualTrees.Header` file in `TotalWidth` method, especially on:
```delphi
    if not (coVisible in Items[LastColumn].Options) then
      LastColumn := GetPreviousVisibleColumn(LastColumn);
```
But have same position, so still return same column.

Checked on the last version of VST.
Sample project attached.
[Projects.zip](https://github.com/JAM-Software/Virtual-TreeView/files/13379332/Projects.zip)

**Comment by joachimmarder** [comment]
Please respect our guidelines on the project homepage for submitting bugs. Please include your version of Virtual TreeView and Delphi, and attach a sample compiling project as ZIP to your report, or instructions how to modify one of the included sample projects to reproduce the problem. If only small changes are required, a description is sufficient how a demo projects needs to be changed in order to replicate the bug. If you already have a solution, please supply a patch file.

**Comment by Dzyszla** [comment]
> Please respect our guidelines on the project homepage for submitting bugs. Please include your version of Virtual TreeView and Delphi, and attach a sample compiling project as ZIP to your report, or instructions how to modify one of the included sample projects to reproduce the problem. If only small changes are required, a description is sufficient how a demo projects needs to be changed in order to replicate the bug. If you already have a solution, please supply a patch file.

Done (issue has been edited). I think this is very simple case. Only VST with 3  columns and my code - that's all.

**Comment by joachimmarder** [comment]
@Dzyszla : Please check if the recent commit fixes the problem for you.

**Comment by Dzyszla** [comment]
Problem was fixed, but... Now no change column positions :( Try change (manually, run-time) column c2 to position 0 (first) and press button - no error, but nothing change (c2 will back to second place).

**Comment by joachimmarder** [comment]
You are right. Could you give it another try?

By the way, it will always work if you don't use `BeginUpdate/EndUpdate` on the columns.

**Comment by Dzyszla** [comment]
In first repeat-until loop iteration for positions [0, 1, 1] second column got position 0. Then in second iteration sets to [-1 (without sign), 0, 1] so... Access Violation :)

I think it works this way: write all position into array, sort it by values (positions) and rewrite from index to position.

Maybe something like this:
```delphi
procedure TVirtualTreeColumns.FixPositions;
// Fixes column positions after loading from DFM or Bidi mode change.
var
  lPositions: System.Generics.Collections.TList<TVirtualTreeColumn>;
  I : Integer;
begin
  lPositions := TList<TVirtualTreeColumn>.Create;
  try
	lPositions.Capacity := Count;
	  for i := 0 to Count - 1 do
		  lPositions.Add(Items[i]);
	lPositions.Sort(
		System.Generics.Defaults.TComparer<TVirtualTreeColumn>.Construct(
			function(const aLeft, aRight: TVirtualTreeColumn): Integer
			begin
				Result := aLeft.Position - aRight.Position;
				if Result = 0 then
					Result := aLeft.Index - aRight.Index; // for not randomize order with same position
			end));
	for i := 0 to Count - 1 do
	begin
		lPositions[i].FPosition := i;
		FPositionToIndex[i] := lPositions[i].Index; // (?)
	end;
  finally
	  lPositions.Free;
  end;

  FNeedPositionsFix := False;
  UpdatePositions(True);
end;
```

----

I have one more ask - it's possible, when official fix will be available, to apply it to the 7.5.x sable version as fix the bug? Or 8.0 will be soon?

**Comment by joachimmarder** [comment]
> Or 8.0 will be soon?

Yes. Actually, I wanted to release V8 this week. This is the last remaining issue for V8.0. 
See: https://github.com/JAM-Software/Virtual-TreeView/milestone/15

**Comment by joachimmarder** [comment]
You should use `BeginUpdate()` / `EndUpdate()` only in case you are setting the entire columns with consistent data. Otherwise just omit it. I will check back to this issue in V8.1.

**Comment by Dzyszla** [comment]
Everywhere using this two commands are used for block unnecessary operations during changing several items of list, so this will works in all case ;) I keep waiting for fix.

**Comment by tecmatia-dp** [comment]
I was expieriencing a lot of problems with version 8.0 in my project, with a tree with 45 columns that the user can move, hide, save multiple views.  In some cases, resulting in exceptions, columns that are not displayed, different widths in header and content, etc.
I found that problem is repeated Position, including design-time. 
I tried to fix positions in my own code, outside the library, but no success.
Finding this open issue, I have seen the way, and fixed method TVirtualTreeColumns.FixPositions. Based on @Dzyszla proposal, this is my final code, that is working fine at runtime:

```
procedure TVirtualTreeColumns.FixPositions;
// Fixes column positions after loading from DFM or Bidi mode change.
var
  LColumnsByPos: TList<TVirtualTreeColumn>;
  I: Integer;
begin
  LColumnsByPos := TList<TVirtualTreeColumn>.Create;
  try
    LColumnsByPos.Capacity := Self.Count;
    for I := 0 to Self.Count-1 do
      LColumnsByPos.Add(Items[I]);

    LColumnsByPos.Sort(
      TComparer<TVirtualTreeColumn>.Construct(
        function(const A, B: TVirtualTreeColumn): Integer
        begin
          Result := CompareValue(A.Position, B.Position);
          if Result = 0 then
            Result := CompareValue(A.Index, B.Index);
        end)
    );

    for I := 0 to LColumnsByPos.Count-1 do
    begin
      LColumnsByPos[I].FPosition := I;
      Self.FPositionToIndex[I] := LColumnsByPos[I].Index;
    end;

  finally
	  LColumnsByPos.Free;
  end;

  FNeedPositionsFix := False;
  UpdatePositions(True);
end;
```
It's the same idea, but using `CompareValue` in Sort (easy to read), and fixing the assignment to the array `FPositionToIndex`.
Thank you @Dzyszla

**Comment by Dzyszla** [comment]
And thank's You too, @tecmatia-dp 
I see, that line with question mark in my code is a little bit different. Maybe that's why my code wasn't working properly always.
Now waiting for official release with fix :)

**Comment by tecmatia-dp** [comment]
Oops, @Dzyszla, checking that line again, I think the correct one is yours, mine always assign Position=I to Index=I...
Correct is `FPositionToIndex[I] := LColumnsByPos[I].Index`
But, why wasn't it working properly always for you?
Now, I think it is working for me, apparently.

Edited my previous code, to fix that line, and change var names to be more easy to read and understand the algorithm (yours).

**Comment by joachimmarder** [comment]
Thanks for the code, which I just published as part of official release V8.0.1

**Comment by Dzyszla** [comment]
@tecmatia-dp 

> But, why wasn't it working properly always for you? Now, I think it is working for me, apparently.

I don't know :( Had some problem with correct changing position... I don't remember now, what exactly case. I try again now after update to 8.0.1, but hope now is all fine :)
